src/main/java/uk/org/okapibarcode/backend/Pdf417AutoStructuredAppend.java

```diff
@@ -0,0 +1,154 @@
+     *
+     * @param data     the data to encode
+     * @param template the PDF417 symbol template which will be used for all created
+     *                 symbols. The following properties will be ignored from the template:
+     *                 <ul>
+     *                      <li>{@link Pdf417#setContent(String)}</li>
+     *                      <li>{@link Pdf417#setContent(byte[])}</li>
+     *                      <li>{@link Pdf417#setStructuredAppendTotal(int)}</li>
+     *                      <li>{@link Pdf417#setStructuredAppendPosition(int)}</li>
+     *                 </ul>
+     * @throws OkapiException if no data or data is invalid
+     */
+    public static List<Pdf417> createStructuredAppendSymbols(String data, Pdf417 template) {
+        List<String> dataList = splitData(data, template);
+        return plotStructuredAppendSymbols(dataList, template);
+    }
+
+    /**
+     * Overloaded method of {@link #createStructuredAppendSymbols(String, Pdf417)} which
+     * accepts binary data and passes it to {@link Pdf417#setContent(byte[])}.
+     *
+     * <p><b>NOTE:</b>See linked method above for details.
+     */
+    public static List<Pdf417> createStructuredAppendSymbols(byte[] data, Pdf417 template) {
+        return createStructuredAppendSymbols(new String(data, ISO_8859_1), template);
+    }
+
+    private static List<String> splitData(String data, Pdf417 template) {
+        List<String> dataList = new ArrayList<>();
+        String remainingData = data;
+        while (!remainingData.isEmpty()) {
+            int low = 0;
+            int high = remainingData.length();
+
+            while (low <= high) {
+                int mid = low + high >>> 1;
+                int currentLength = Math.min(mid, remainingData.length());
+                String currentData = remainingData.substring(0, currentLength);
+                boolean fit = fits(currentData, template, false);
+                if (fit && mid < remainingData.length() &&
+                    fits(remainingData.substring(0, currentLength + 1), template, false)) {
+                    low = mid + 1;
+                } else {
+                    if (fit && currentLength == mid) {
+                        dataList.add(currentData);
+                        remainingData = remainingData.substring(mid);
+                    }
+                    high = mid - 1;
+                }
+            }
+        }
+        if (!dataList.isEmpty()) {
+            String lastData = dataList.get(dataList.size() - 1);
+            if (!fits(lastData, template, true)) {
+                int endIndex = lastData.length() - 1;
+                dataList.set(dataList.size() - 1, lastData.substring(0, endIndex));
+                dataList.add(lastData.substring(endIndex));
+            }
+        }
+        return dataList;
+    }
+
+    private static boolean fits(String data, Pdf417 template, boolean last) {
+        Pdf417 symbol = new Pdf417() {
```

Comment:
Instead of instantiating here, we could even instantiate out of the loops, and reuse the one instance for all the tests. Do you think that would be safe?

Reply:
I have changed it like that, when passing in huge amount of data, instantiating testing symbols showed up in profiling. Let me know in case you see issues with this approach.